File: iOS/PanoAudioChat/PanoAudioChatSDK/Source/Sections/AudioRoom/View/Player/PanoAudioPlayerView.m
//
//  PanoAudioPlayerView.m
//  PanoAudioChatSDK
//
//  Created by pano on 2021/7/19.
//

#import "PanoAudioPlayerView.h"

@implementation PanoAudioPlayerView
{
    UIView *contentView;

    UIButton *nextBtn;

    UIImageView *icon;
}

- (void)initViews {
    contentView = [UIView new];
    contentView.layer.masksToBounds = true;
    contentView.layer.cornerRadius = 19;
    contentView.backgroundColor = [UIColor pano_colorWithHexString:@"#577096"];
    [self addSubview:contentView];
    
    icon = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"room_player_music"]];
    [contentView addSubview:icon];
    
    playBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    [playBtn setImage:[UIImage imageNamed:@"room_player_pause"] forState:UIControlStateNormal];
    [playBtn setImage:[UIImage imageNamed:@"room_player_play"] forState:UIControlStateSelected];
    playBtn.selected = true;
    [playBtn addTarget:self action:@selector(didStartPlayAction:) forControlEvents:UIControlEventTouchUpInside];
    [contentView addSubview:playBtn];
    
    nextBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    [nextBtn setImage:[UIImage imageNamed:@"room_player_next"] forState:UIControlStateNormal];
    [nextBtn setImage:[UIImage imageNamed:@"room_player_next"] forState:UIControlStateSelected];
    [nextBtn addTarget:self action:@selector(didChooseNextAction) forControlEvents:UIControlEventTouchUpInside];
    [contentView addSubview:nextBtn];
  
    [self addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(showMoreAction)]];
}

- (void)initConstraints {
    [contentView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.width.mas_equalTo(120);
        make.edges.mas_equalTo(0);
    }];
    
    [icon mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.mas_equalTo(0);
        make.left.mas_equalTo(0).mas_offset(3);
        make.size.mas_equalTo(CGSizeMake(30, 30));
    }];
    
    [playBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(icon.mas_right).offset(5);
        make.centerY.mas_equalTo(0);
        make.size.mas_equalTo(CGSizeMake(30, 30));
    }];
    
    [nextBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(playBtn.mas_right).offset(5);
        make.centerY.mas_equalTo(0);
        make.size.mas_equalTo(CGSizeMake(30, 30));
    }];
}

- (void)didStartPlayAction:(BOOL)pause {
    playBtn.selected = !playBtn.selected;
    [_delegate didStartPlayAction:playBtn.isSelected];
}

- (void)didChooseNextAction {
    playBtn.selected = false;
    [_delegate didChooseNextAction];
}

- (void)showMoreAction {
    [_delegate didShowMoreAction];
}


@end

File: iOS/PanoAudioChat/PanoAudioChat/Classes/Sections/AudioRoom/View/Player/PanoAudioPlayerView.swift
//
//  PanoAudioPlayerView.swift
//  PanoAudioChat
//
//  Created by wenhua.zhang on 2020/11/4.
//

import UIKit

protocol PanoAudioPlayerDelegate : NSObjectProtocol {
    
    func didChooseNextAction()
    
    func didStartPlayAction(pause: Bool)
    
    func didShowMoreAction()
}

class PanoAudioPlayerView: PanoBaseView {

    weak var delegate: PanoAudioPlayerDelegate?
    
    var contentView: UIView!
    
    var playBtn: UIButton!
    
    var nextBtn: UIButton!
    
    var audioName: UILabel!
    
    var audioStateLabel: UILabel!
    
    var icon: UIImageView!
    
    override func initViews() {
        
        contentView = UIView()
        contentView.layer.masksToBounds = true
        contentView.layer.cornerRadius = 19
        contentView.backgroundColor = UIColor("#577096")
        self.addSubview(contentView)
        
        icon = UIImageView(image: UIImage(named: "room_player_music"))
        contentView.addSubview(icon)
        
        playBtn = UIButton()
        playBtn.setImage(UIImage(named: "room_player_pause"), for: .normal)
        playBtn.setImage(UIImage(named: "room_player_play"), for: .selected)
        playBtn.isSelected = true
        playBtn.addTarget(self, action: #selector(didStartPlayAction(pause:)), for: .touchUpInside)
        self.contentView.addSubview(playBtn)
        
        nextBtn = UIButton()
        nextBtn.setImage(UIImage(named: "room_player_next"), for: .normal)
        nextBtn.setImage(UIImage(named: "room_player_next"), for: .highlighted)
        nextBtn.addTarget(self, action: #selector(didChooseNextAction), for: .touchUpInside)
        self.contentView.addSubview(nextBtn)
        
        self.addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(showMoreAction)))
        
    }
    
    override func initConstraints() {
        contentView.snp.makeConstraints { (make) in
            make.width.equalTo(120)
            make.edges.equalToSuperview()
        }
        icon.snp.makeConstraints { (make) in
            make.centerY.equalToSuperview()
            make.left.equalToSuperview().offset(3)
            make.size.equalTo(CGSize(width: 30, height: 30))
        }
//        audioName.snp.makeConstraints { (make) in
//            make.left.equalTo(icon.snp.right).offset(3)
//            make.top.equalToSuperview()
//            make.width.equalTo(40)
//            make.height.equalTo(19)
//        }
//        audioStateLabel.snp.makeConstraints { (make) in
//            make.left.equalTo(audioName.snp.right).offset(3)
//            make.top.equalToSuperview()
//            make.width.equalTo(40)
//            make.centerY.equalTo(audioName)
//        }
        playBtn.snp.makeConstraints { (make) in
            make.left.equalTo(icon.snp.right).offset(5)
            make.centerY.equalToSuperview()
            make.size.equalTo(CGSize(width: 30, height: 30))
        }
        
        nextBtn.snp.makeConstraints { (make) in
            make.left.equalTo(playBtn.snp.right).offset(5)
            make.centerY.equalToSuperview()
            make.size.equalTo(CGSize(width: 30, height: 30))
        }
    }
    
    @objc func didStartPlayAction(pause: Bool) {
        playBtn.isSelected = !playBtn.isSelected
        delegate?.didStartPlayAction(pause: playBtn.isSelected)
    }
    
    @objc func didChooseNextAction() {
        playBtn.isSelected = false
        delegate?.didChooseNextAction()
    }
    
    @objc func showMoreAction() {
        delegate?.didShowMoreAction()
    }
}
